oak/client/sgx_application_client.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "oak/client/sgx_application_client.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "asylo/grpc/auth/enclave_channel_credentials.h"
+#include "asylo/grpc/auth/enclave_credentials_options.h"
+#include "asylo/grpc/auth/null_credentials_options.h"
+// TODO: Uncomment when Asylo will release remote attestation.
+//#include "asylo/grpc/auth/sgx_remote_credentials_options.h"
+#include "asylo/identity/descriptions.h"
+#include "asylo/identity/enclave_assertion_authority_configs.h"
+#include "asylo/identity/init.h"
+#include "asylo/identity/sgx/sgx_identity_util.h"
+#include "asylo/identity/util/sha256_hash.pb.h"
+#include "asylo/util/logging.h"
+#include "asylo/util/status.h"
+#include "asylo/util/statusor.h"
+#include "oak/client/authorization_bearer_token_metadata.h"
+#include "oak/client/policy_metadata.h"
+#include "oak/common/nonce_generator.h"
+#include "oak/common/policy.h"
+
+using asylo::Status;
+using asylo::StatusOr;
+
+constexpr size_t kPerChannelNonceSizeBytes = 32;
+// Debug MRSIGNER value derived from the SGX test key (will change after changing SVN and PRODID).
+// https://github.com/google/asylo/blob/088ea3490dd4579655bd5b65b0e31fe18de7f6dd/asylo/distrib/sgx_x86_64/linux_sgx_2_6.patch#L5481
+const char* kDebugMrSigner = "83d719e77deaca1470f6baf62a4d774303c899db69020f9c70ee1dfc08c7ce9e";
+// Debug CPUSVN value.
+// This value should be provided by EREPORT and it reflects the microcode update version.
+const char* kDebugCpuSvn = "0000000000000000";
+
+// Parses a hexademical hash string from `hash_string` into `Sha256HashProto` in `hash`.
+// TODO: Use same function from Asylo, when it becomes public.
+Status Sha256HashFromHexString(const std::string& hash_string, asylo::Sha256HashProto* hash) {
```

Comment:
The comment should mention that it expects a string of 64 bytes. Although it seems odd to couple this check to the conversion. Personally I would let it convert any hex string to bytes (like `absl::HexStringToBytes`, but with validity checks and returning `StatusOr<string>` or similar), then have the caller check the length of that output bytes to be 32 (or whatever), then finally put that in the proto. You can definitely have a function `SetValueInProto` that handles all this logic, if you don't want to have separate checks in `CreateSgxIdentityExpectation`. WDYT?

Reply:
Done